### PR TITLE
Extend Ordering Equivalence Support

### PR DESCRIPTION
**Pull request by mustafasrepo**
# Which issue does this PR close?

Closes #.

# Rationale for this change
Currently ordering equivalence properties cannot be propagated through ExecutionPlan. However, for executors that maintain its input order, we should be able to propagate ordering equivalence properties. This PR implements ordering equivalence properties for executors that maintains ordering. For most of the executors, this is trivial (such as `FilterExec`).
However, for `SortMergeJoinExec` and `HashJoinExec` updating ordering equivalence involves non-trivial implementation. Reviewers can focus on these executors.

Also during tests for propagating ordering equivalence properties through executors. I found a bug in how Projection handles ordering equivalence and equivalence. This PR fixes it also.

# What changes are included in this PR?

# Are these changes tested?

Yes new tests are added to `window.slt` and `joins.slt` file

# Are there any user-facing changes?

### CODE REVIEW

**Review by metesynnada — COMMENTED** [review]
LGTM!

File: datafusion/core/src/physical_plan/joins/sort_merge_join.rs [hunk 1]
```diff
@@ -295,6 +321,63 @@ impl ExecutionPlan for SortMergeJoinExec {
...
+                // Since left side is the stream side for this `SortMergeJoin` implementation,
+                // global ordering of the left table is preserved at the output. Hence, left
+                // side ordering equivalences are still valid.
+                new_properties.extend(left_oeq_properties.classes().iter().cloned());
+                if let Some(output_ordering) = &self.output_ordering {
+                    // We need to add ordering equivalence properties of right
+                    // table as append to the lexicographical ordering to the existing ordering.
+                    // For instance, if right table ordering equivalence contains b ASC
+                    // And output ordering of the SortMergeJoin is a ASC, We should add
+                    // ordering equivalence `b ASC` for right table as `a ASC, b ASC` for `SortMergeJoinExec`.
```
Comment (ozankabak) [inline datafusion/core/src/physical_plan/joins/sort_merge_join.rs:340]: Let's clarify this comment.

File: datafusion/core/src/physical_plan/projection.rs [hunk 2]
```diff
@@ -299,6 +316,30 @@ impl ExecutionPlan for ProjectionExec {
     }
 }
 
+fn update_ordering_equivalence_with_cast(
```
Comment (ozankabak) [inline datafusion/core/src/physical_plan/projection.rs:319]: This function needs a docstring.

**Review by ozankabak — COMMENTED** [review]
Two minor comment-related issues, let's fix the cast issue and review again.

**Review by ozankabak — APPROVED** [review]

**Comment by mustafasrepo** [comment]
Merged upstream

**Pull request closed without merging**